### PR TITLE
Accessible ui select

### PR DESCRIPTION
**Pull request by 0xF013**
[`* `#17790` #17790`](http://community.openproject.org/work_packages/17790)
[`* `#17787` #17787`](http://community.openproject.org/work_packages/17787)
[`* `#17789` #17789`](http://community.openproject.org/work_packages/17789)

### CODE REVIEW

**Comment by TeatroIO** [comment]
I've prepared a stage. <a href='http://0xf013-accessible-ui-select.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Click to open.</a>

File: frontend/bower.json [hunk 1]
```diff
@@ -26,7 +26,7 @@
     "angular-busy": "~4.1.1",
     "hyperagent": "manwithtwowatches/hyperagent#v0.4.2",
     "lodash": "~2.4.1",
-    "ui-select": "angular-ui/ui-select#~0.9.5"
+    "ui-select": "0xf013/ui-select#add-combobox-acessibility"
```
Comment (myabc) [inline frontend/bower.json]: @0xF013 did you change the branch name?

Comment (0xF013) [inline frontend/bower.json]: @myabc yeah, no. Will change it

**Comment by ulferts** [comment]
http://community.openproject.org/work_packages/17790

**Comment by ulferts** [comment]
http://community.openproject.org/work_packages/17787

**Comment by ulferts** [comment]
http://community.openproject.org/work_packages/17789

**Comment by 0xF013** [comment]
@myabc @ulferts I switched to the latest build (from my repository) since the ui-select rep owner didn't build the sources. It's ready to be reviewed

**Comment by ulferts** [comment]
@0xF013, can you please rebase?

**Comment by 0xF013** [comment]
@ulferts done

File: config/locales/js-de.yml [hunk 2]
```diff
@@ -386,9 +386,9 @@ de:
       delete: Beziehung löschen
     inplace:
       button_edit: "%{attribute} bearbeiten"
-      button_save: "Speichern"
-      button_save_and_send: "Speichern mit E-Mail-Benachrichtigung"
-      button_cancel: "Abbrechen"
+      button_save: "%{attribute} Speichern"
```
Comment (myabc) [inline config/locales/js-de.yml]: @0xF013 sorry for just noticing this. To match with the English variant, I guess there should be a colon after the attribute name.

**Comment by 0xF013** [comment]
@myabc fixed semicolon

**Comment by myabc** [comment]
@0xF013 danke

**Comment by ulferts** [comment]
@0xF013 we will probably have to improve drop-downs with empty values for selection (e.g. assignee and responsible). Nothing except "1 of X" is read out when hovering over such an empty field.

**Comment by 0xF013** [comment]
@ulferts ui-select has a setting for being able to clear the input, but it's an inaccessible span, this is why I am simply adding an empty option in the inplace editor. We can add a text to that empty option that says 'No value'.

**Comment by 0xF013** [comment]
@myabc @ulferts latest commit should add a readable label to the null option as well as read the current value when the select is activated

**Comment by RobinWagner** [comment]
@0xF013 I took a brief look at the fix. The options in the selection menu are properly pronounced. Is it possible to also say how many options there are? (if there are 5 options: "there are 5 work packages matching the entered value. Use keyboard arrows to select the right option." Then on switching via arrows: "Option 1: #23232 Example work package" etc.)
(If this is a lot of work, just let me know, then I'll talk to Mrs. Ruth-Janneck to see if the suggest implementation is also sufficient)

This is probably not part of this PR but another problem I see is with confirming the selected option: The user has to tab back multiple times to get to the "Add relation" button. When focusing the Button the selected work package is not read by the screen reader, so a blind user actually does not know what work package he is adding. Ideally, the tab order would be reversed, so that a user first selects a work package from the UI-select, then tabs to "Add relation" (and ideally gets a reminder which work package will be added).

**Comment by 0xF013** [comment]
@RobinWagner Please check the accessibility on the status or responsibles. The issues with filtering are known and I am looking for a solution. I will address the relations tab in a separate PR

**Comment by 0xF013** [comment]
@RobinWagner I have pushed a new commit that deals with live filtering

**Comment by 0xF013** [comment]
@RobinWagner the number of items and the current index is read out by the reader standardly. I don't know how to make it also read 'Use keyboard arrows to select the right option' . We'll need to think on how to help users about arrows

**Comment by 0xF013** [comment]
@RobinWagner about ' "Option 1: #23232 Example work package" ', I think the reader tells something like '2 out on 11'. We could enhance the text with an 'Option x' prefix

**Comment by 0xF013** [comment]
@RobinWagner again on arrows: I think it's the default usecase for dropdowns and people are aware, also the ui.select now mimics the dropdown metadata and is presented to the reader as a regular dropdown plus filtering

**Comment by ulferts** [comment]
- After pressing on the "Edit" icon, the drop down is presented and the initially selected value (e.g. "Closed") is read along with it's positioning in the list. That is good. I can now make my selection via the up/down-buttons and press enter (or tab) to confirm the selection. When using the "Enter" key, the focus is then put on an item which the screen reader reads out as e.g. "Status, ended focus ..."(?). But when I now want to change my selection, I cannot use the "Enter" key again because it will already change the value by selecting the first value in the list. I have to use the up/down keys to reopen the dropdown without changing the value. I have no clue whether that is acceptable, but when closing my eyes, I found it rather confusing.
- When I use the autocompleter functionality to narrow down the list of available values, the feedback I receive is hard to understand. When I type the screen reader reads out the characters I type. It also reads out, when the first value changes (and only then). But it doesn't tell me how many values are still in the list. This is especially annoying when there is only one (or no) value in the list.

But as a whole, this is looking pretty good. Needs merge, though.

**Comment by ulferts** [comment]
@0xF013 is there anything you can do on about the points I noted?

**Comment by 0xF013** [comment]
@ulferts apparently the fix I did for the second issue works with voice over but not with chrome vox. I will take a look tomorrow at that and the focus issue

**Comment by 0xF013** [comment]
@ulferts pushed a possible fix, at least is works on my setup. please remove the frontend/bower_components and run a bower install and a gulp webpack

**Comment by ulferts** [comment]
@0xF013 this really seems to work :+1:

**Pull request merged**